### ISSUE TITLE
Out of memory on android in DefaultMp4Builder build step

### ISSUE THREAD

**Issue by rbolyard** [sannies/mp4parser#76]
Trace:
 java.io.IOException: mmap failed: ENOMEM (Out of memory)
            at java.nio.MemoryBlock.mmap(MemoryBlock.java:125)
            at java.nio.FileChannelImpl.map(FileChannelImpl.java:257)
            at com.demo.libraries.AndroidUriDataSource.map(AndroidUriDataSource.java:90)
            at com.googlecode.mp4parser.BasicContainer.getByteBuffer(BasicContainer.java:223)
            at com.googlecode.mp4parser.authoring.samples.DefaultMp4SampleList.get(DefaultMp4SampleList.java:183)
            at com.googlecode.mp4parser.authoring.samples.DefaultMp4SampleList.get(DefaultMp4SampleList.java:16)
            at com.coremedia.iso.boxes.mdat.SampleList.get(SampleList.java:42)
            at com.coremedia.iso.boxes.mdat.SampleList.get(SampleList.java:22)
            at java.util.AbstractList$SubAbstractList.get(AbstractList.java:292)
            at com.googlecode.mp4parser.authoring.builder.DefaultMp4Builder.build(DefaultMp4Builder.java:100)
. . .

Caused by: android.system.ErrnoException: mmap failed: ENOMEM (Out of memory)
            at libcore.io.Posix.mmap(Native Method)
            at libcore.io.ForwardingOs.mmap(ForwardingOs.java:111)
            at java.nio.MemoryBlock.mmap(MemoryBlock.java:122)
            at java.nio.FileChannelImpl.map(FileChannelImpl.java:257)
            at com.demo.libraries.AndroidUriDataSource.map(AndroidUriDataSource.java:90)
            at com.googlecode.mp4parser.BasicContainer.getByteBuffer(BasicContainer.java:223)
            at com.googlecode.mp4parser.authoring.samples.DefaultMp4SampleList.get(DefaultMp4SampleList.java:183)
            at com.googlecode.mp4parser.authoring.samples.DefaultMp4SampleList.get(DefaultMp4SampleList.java:16)
            at com.coremedia.iso.boxes.mdat.SampleList.get(SampleList.java:42)
            at com.coremedia.iso.boxes.mdat.SampleList.get(SampleList.java:22)
            at java.util.AbstractList$SubAbstractList.get(AbstractList.java:292)
            at com.googlecode.mp4parser.authoring.builder.DefaultMp4Builder.build(DefaultMp4Builder.java:100)
. . .

Use case:
Trimming the audio/video tracks in a longer 1080p video recorded on an android phone down to around 5+ min seems to require more than the 512MB that the android app is allotted.

Here's what I've found:
DefaultMp4Builder.build() starts by looping through all samples, getting their sizes.  Each call to samples.get(i) is going through DefaultMp4SampleList.get().  That code retrieves each chunk as needed and stores the chunk in a cache.  This cache is never cleaned up.  The android app runs out of memory around 512MB.

I did see that there were code changes back in November that added a MAX_MAP_SIZE.  However, this appears to only check whether an individual chunk is larger than MAX_MAP_SIZE and not the whole cache.  So this logic is never triggered.

I dug deeper and found that the cache used to be a SoftReference array.  I experimented with using a SoftReference array again (keeping all newer code changes since its removal) and found it to work perfectly for my case.  This amounts to changing 4 lines of code.

What was the reasoning for moving away from SoftReference?  Does it make sense to go back to SoftReference?

Let me know if you'd like any additional information.

**Comment by sannies** [comment]
Hmm. It shouldn't cost much memory as the cache is not filled with data structure on the heap. Typically we have ByteBuffers in the cache that have been create with a memmap operation - as in your case. These datastructure reside in the virtual memory and the operating system decide whether it's in the memory or on disk. It shouldn't increase the JVM heap! 
I'm a bit puzzled here. 
Does it the same operation work on a PC? If it also fails there it might be easier to diagnose. 
Do you perhaps have an example I can run on my Galaxy Note 3? Perhaps it is even a device dependent issue

**Comment by rbolyard** [comment]
Ah, figured out the confusion on my end.  That stack trace I posted here was from when we used to use memory mapping.  When we did, Android debug tools did not show any memory usage during DefaultMp4Builder.build, but we still encountered OOM exceptions (and couldn't do anything about it).  This was encountered on multiple devices consistently.

Since that time, we changed the allocations to explicitly use the JVM heap instead, which allowed for cleanup rather than crashing.  So that explains the use of the heap.  As long as we use the heap, we'll need the cache to garbage collect.  SoftReference seems like the best fix for that.

As for why memory mapping was causing OOM on Android, I'm not sure.  That issue aside, is there any reason not to use SoftReference in DefaultMp4SampleList?  It seems like it will work just fine for either heap allocations or memory mapped reads.  So I'd still suggest that fix.  I can submit a pull request if you'd like.
